### PR TITLE
cmdline interface + time benchmark to LTR module

### PR DESCRIPTION
**Pull request by stephaniewhoo**
As mentioned in #1387 #1386, it has simple cmd line interface for debugging purpose. Also include some debug operations for concurrent problem

### CODE REVIEW

**Comment by lintool** [comment]
@nsndimt can you take a look?

@stephaniewhoo any insights on what's fast and what's slow?

**Comment by stephaniewhoo** [comment]
> @nsndimt can you take a look?
> 
> @stephaniewhoo any insights on what's fast and what's slow?

Here's a sample run.
```
AvgICTF takes 6.90 s account for 2.08%
AvgIDF takes 1.19 s account for 0.36%
BM25 takes 12.25 s account for 3.69%
DocSize takes 0.59 s account for 0.18%
MatchingTermCount takes 5.80 s account for 1.74%
PMI takes 149.78 s account for 45.05%
QueryLength takes 0.53 s account for 0.16%
AvgSCQ takes 1.22 s account for 0.37%
SCS takes 0.85 s account for 0.25%
SumMatchingTF takes 7.36 s account for 2.21%
TFIDF takes 7.27 s account for 2.19%
UniqueQueryTerms takes 1.00 s account for 0.30%
UnorderedSequentialPairs3 takes 28.45 s account for 8.56%
UnorderedSequentialPairs5 takes 25.54 s account for 7.68%
UnorderedSequentialPairs8 takes 26.51 s account for 7.97%
OrderedSequentialPairs3 takes 19.21 s account for 5.78%
OrderedSequentialPairs5 takes 18.73 s account for 5.63%
OrderedSequentialPairs8 takes 19.29 s account for 5.80%
```
Still PMI takes dominant time.

**Comment by lintool** [comment]
And what's the impact of PMI on effectiveness? I.e., what's MRR if we take it out?

**Comment by stephaniewhoo** [comment]
> And what's the impact of PMI on effectiveness? I.e., what's MRR if we take it out?
It's a bit weird...
While commenting out PMI is 
```
#####################
MRR @10: 0.19582222222222223
QueriesRanked: 500
#####################
```
and with PMI
```
#####################
MRR @10: 0.1860611111111111
QueriesRanked: 500
#####################
```
@nsndimt had same result when commenting out PMI I think, not sure if we have same MRR number with PMI

**Comment by lintool** [comment]
So, good reason to ditch the feature!

**Comment by nsndimt** [comment]
same MRR after adding PMI back
`MRR @10: 0.1860611111111111`

similar benchmark result with 20 threads
```
AvgICTF takes 13.19 s account for 2.12%
AvgIDF takes 2.14 s account for 0.34%
BM25 takes 20.83 s account for 3.34%
DocSize takes 0.90 s account for 0.14%
MatchingTermCount takes 11.22 s account for 1.80%
PMI takes 265.01 s account for 42.52%
QueryLength takes 0.83 s account for 0.13%
AvgSCQ takes 2.06 s account for 0.33%
SCS takes 1.52 s account for 0.24%
SumMatchingTF takes 13.46 s account for 2.16%
TFIDF takes 13.85 s account for 2.22%
UniqueQueryTerms takes 1.77 s account for 0.28%
UnorderedSequentialPairs3 takes 53.50 s account for 8.58%
UnorderedSequentialPairs5 takes 51.61 s account for 8.28%
UnorderedSequentialPairs8 takes 54.88 s account for 8.81%
OrderedSequentialPairs3 takes 38.56 s account for 6.19%
OrderedSequentialPairs5 takes 38.29 s account for 6.14%
OrderedSequentialPairs8 takes 39.61 s account for 6.36%
The program takes 65.23 s where the overhead takes 1.03 s
```

File: src/main/java/io/anserini/ltr/FeatureExtractorCli.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.anserini.ltr.feature.OrderedSequentialPairsFeatureExtractor;
+import io.anserini.ltr.feature.UnorderedSequentialPairsFeatureExtractor;
+import io.anserini.ltr.feature.base.*;
+import org.kohsuke.args4j.CmdLineException;
+import org.kohsuke.args4j.CmdLineParser;
+import org.kohsuke.args4j.Option;
+
+import java.io.*;
```
Comment (lintool) [inline src/main/java/io/anserini/ltr/FeatureExtractorCli.java:12]: please expand out *

File: src/main/java/io/anserini/ltr/FeatureExtractorCli.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import java.util.ArrayList;
+import java.util.concurrent.ExecutionException;
+
+public class FeatureExtractorCli {
```
Comment (lintool) [inline src/main/java/io/anserini/ltr/FeatureExtractorCli.java:16]: we use two space indent, so please change throughout?

File: src/main/java/io/anserini/ltr/FeatureExtractorCli.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        utils.add(new OrderedSequentialPairsFeatureExtractor(3));
+        utils.add(new OrderedSequentialPairsFeatureExtractor(5));
+        utils.add(new OrderedSequentialPairsFeatureExtractor(8));
+
+
+
+        File file = new File(cmdArgs.jsonFile);
+        BufferedReader reader = new BufferedReader(new InputStreamReader(new FileInputStream(file)));
+        String line;
+        ArrayList<String> qids = new ArrayList<>();
```
Comment (lintool) [inline src/main/java/io/anserini/ltr/FeatureExtractorCli.java:67]: just `List<String>`

Comment (lintool) [reply src/main/java/io/anserini/ltr/FeatureExtractorCli.java:67]: Similar comments below... don't expose implementations unless absolutely necessary.

**Review by lintool — COMMENTED** [review]
light initial CR.

**Comment by stephaniewhoo** [comment]
what should I do to pass code coverage check...

File: src/main/java/io/anserini/ltr/FeatureExtractorUtils.java [hunk 4]
```diff
@@ -61,6 +66,13 @@ public FeatureExtractorUtils add(FeatureExtractor extractor) {
         return this;
     }
 
+    public ArrayList<String> list() {
```
Comment (lintool) [inline src/main/java/io/anserini/ltr/FeatureExtractorUtils.java:69]: this still appears to be 4-space indent?

**Comment by lintool** [comment]
It'd be pretty difficult to write test cases for this... @nsndimt any testing ideas?

If we don't have any reasonable ideas, then we can take the hit - I can override and merge when everything is sorted ot.

**Comment by nsndimt** [comment]
I cannot think of a quick and good testing plan for this cmdline interface. I think we skip this for now.

**Review by lintool — APPROVED** [review]

**Pull request merged**